arch/arm/core/swap.S

```diff
@@ -176,12 +177,24 @@ _thread_irq_disabled:
 #endif /* CONFIG_MPU_STACK_GUARD */
 
 #ifdef CONFIG_USERSPACE
+    /* restore mode */
+    ldr r0, [r2, #_thread_offset_to_mode]
+    mrs r3, CONTROL
+    bic r3, #1
+    orr r3, r0
+    msr CONTROL, r3
+
     /* r2 contains k_thread */
     add r0, r2, #0
     push {r2, lr}
     blx configure_mpu_mem_domain
     pop {r2, lr}
-#endif /* CONFIG_USERSPACE */
+
+    add r0, r2, #0
+    push {r2, lr}
+    blx configure_mpu_user_context
```

Comment:
You didn't start this trend, but wondering if all of these separate calls to configure_mpu_*, each with its own disable/configure/enable sequence, is a good idea.  Performance overhead of disabling/enabling the MPU for each such region?

Reply:
performance overhead is a single register write to enable/disable the regions globally.  This would be equal to or less costly than toggling 1 or more region enables.

Reply:
I thought about collapsing all of the mpu modifications into a single function, but then decided against it mainly because of the separate nature of the features.